bindings/python/pymongoarrow/builder.pyx

```diff
@@ -0,0 +1,94 @@
+# Copyright 2021-present MongoDB, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Cython compiler directives
+# distutils: language=c++
+# cython: language_level=3
+from pyarrow.lib cimport *
+
+import numpy as np
+
+
+cdef class Int32Builder(_Weakrefable):
+    cdef:
+        shared_ptr[CInt32Builder] builder
+
+    def __cinit__(self, MemoryPool memory_pool=None):
+        cdef CMemoryPool* pool = maybe_unbox_memory_pool(memory_pool)
+        self.builder.reset(new CInt32Builder(pool))
+
+    def append(self, value):
+        if value is None or value is np.nan:
+            self.builder.get().AppendNull()
+        elif isinstance(value, int):
+            self.builder.get().Append(value)
+        else:
+            raise TypeError('Int32Builder only accepts integer objects')
+
+    def append_values(self, values):
+        for value in values:
+            self.append(value)
+
+    def finish(self):
+        cdef shared_ptr[CArray] out
+        with nogil:
+            self.builder.get().Finish(&out)
+        return pyarrow_wrap_array(out)
+
+    @property
+    def null_count(self):
+        return self.builder.get().null_count()
+
+    def __len__(self):
+        return self.builder.get().length()
+
+    cdef shared_ptr[CInt32Builder] unwrap(self):
+        return self.builder
+
+
+cdef class Int64Builder(_Weakrefable):
+    cdef:
+        shared_ptr[CInt64Builder] builder
+
+    def __cinit__(self, MemoryPool memory_pool=None):
+        cdef CMemoryPool* pool = maybe_unbox_memory_pool(memory_pool)
+        self.builder.reset(new CInt64Builder(pool))
+
+    def append(self, value):
+        if value is None or value is np.nan:
+            self.builder.get().AppendNull()
+        elif isinstance(value, int):
+            self.builder.get().Append(value)
+        else:
+            raise TypeError('Int64Builder only accepts integer objects')
+
+    def append_values(self, values):
+        for value in values:
+            self.append(value)
+
+    def finish(self):
+        cdef shared_ptr[CArray] out
+        with nogil:
+            self.builder.get().Finish(&out)
+        return pyarrow_wrap_array(out)
+
+    @property
+    def null_count(self):
+        return self.builder.get().null_count()
+
+    def __len__(self):
+        return self.builder.get().length()
+
+    cdef shared_ptr[CInt64Builder] unwrap(self):
+        return self.builder
```

Comment:
Hmm these classes are so similar. There must be a way to share the implementation, right? I found this blog about it https://martinralbrecht.wordpress.com/2017/07/23/adventures-in-cython-templating/

But for now I don't think we should use any of these "templating" options and instead keep it simple.

Reply:
Yeah I tried a bunch of different things for sharing code here since so much of it is common and since Cython does support class inheritance. Unfortunately, the scoping rules enforced on `cdef`-ined variables by Cython are causing problems when I try to share code between these classes. For now I have given up on this front. We can make this more sophisticated after 0.1.0

Reply:
SGTM

Reply:
Update - I did a bit of updating... Some more optimizations will be possible in a subsequent PR.